contrib/xor-blocks/src/main.rs

```diff
@@ -0,0 +1,193 @@
+
+#[derive(Debug)]
+struct Error {
+    msg: String,
+}
+
+impl Error {
+    fn new(msg: String) -> Self {
+        Self { msg }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.msg)
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(value: io::Error) -> Self {
+        Error::new(value.to_string())
+    }
+}
+
+impl From<&str> for Error {
+    fn from(value: &str) -> Self {
+        Error::new(value.to_string())
+    }
+}
+
+fn main() -> Result<(), Error> {
+    let start = Instant::now();
+
+    let data_dir_path = if env::args().len() == 2 {
+        let arg = env::args().nth(1).expect("the arg to exist");
+        if !arg.starts_with("--datadir=") && !arg.starts_with("-datadir=") {
+            return Err(Error::new(format!(
+                "Error parsing command line arguments: Invalid parameter {arg}"
+            )));
+        }
+        let (_, datadir) = arg.split_once("-datadir=").expect("split to be some");
+        datadir.into()
+    } else if env::args().len() == 1 {
+        #[allow(deprecated)]
+        env::home_dir()
+            .expect("to have a home directory")
+            .join(match env::consts::OS {
+                "macos" => "Library/Application Support/Bitcoin",
+                "windows" => "AppData\\Roaming\\Bitcoin",
+                "linux" => ".bitcoin",
+                _ => {
+                    return Err("Unknown OS".into());
+                }
+            })
+    } else {
+        return Err("Invalid number of arguments. Run with `cargo run --release` if using default datadir. Otherwise run with `cargo run --release -- -datadir=<custom datadir>`.".into());
+    };
+
+    let blocks_path = data_dir_path.join("blocks");
+
+    let paths = fs::read_dir(&blocks_path)?
+        .map(|res| res.map(|e| e.path()))
+        .collect::<Result<Vec<_>, io::Error>>()?;
+
+    let key_path = blocks_path.join(XOR_FILE_NAME);
+    let key: [u8; 8] = fs::read(&key_path)
+        .map_err(|_| "No xor.dat file. Make sure you are running Bitcoin Core v28 or higher.")?
+        .try_into()
+        .map_err(|_| "Invalid xor.dat file.")?;
+
+    if key[..4] == [0u8; 4] && key[4..] != [0u8; 4] {
+        return Err(
+            "This tool doesn't work with a non-zero key with 4 bytes of leading zeros".into(),
+        );
+    }
+
+    let key = if key == [0u8; 8] {
+        // The key in the file is zero, so overwrite with a new random one
+        loop {
+            let key: [u8; 8] = RandomState::new().build_hasher().finish().to_le_bytes();
+            // Don't use keys with 4 bytes of leading zeros
+            // They won't let us detect the first 4 bytes of magic in the files
+            if key[..4] == [0u8; 4] {
+                continue;
+            }
+            let mut out_file = File::options().write(true).open(&key_path)?;
+            out_file.write_all(&key)?;
+            out_file.sync_data()?;
```

Comment:
To be atomic, this should create a temporary file first, write it, sync it, and then rename.